### ISSUE TITLE
#2173 breaks flyway extension, and maybe more

### ISSUE THREAD

**Issue by broccolai** [jdbi/jdbi#2208]
After more than one test runs you will get the error that the database is now empty. As far as I can tell its because of this [line](https://github.com/jdbi/jdbi/blob/master/testing/src/main/java/org/jdbi/v3/testing/junit5/JdbiExtension.java#L375).

The flyway extension needs to run every test, so that it can recreate the tables etc. #2173 made it so the extension is removed after every test, meaning it'd only run once.

I would attempt opening a PR for this myself, but I do not want to defeat the original purpose of #2173.

**Comment by hgschmie** [comment]
Hi @broccolai ! Thank you for filing a bug report!

do you have a test case? I modified the JdbiExtensionTest in the test suite to have multiple tests that do all the same thing and with Flyway 9.8.2 I see this output (I added the ---- lines for clarification ----):

```
---- preamble, begin static database extension start ----
[main] INFO de.softwareforge.testing.postgres.embedded.ZonkyIOPostgresLocator - Using PostgreSQL version 13.9.0
[main] INFO de.softwareforge.testing.postgres.embedded.EmbeddedPostgres$NBwrFJfuMd5tpYBZ - started as pid 99133 on port 53001
---- preamble, end static database extension start ----

---- first test, begin spin up database ----
[main] INFO de.softwareforge.testing.postgres.junit5.EmbeddedPgExtension - Connection to jdbc:postgresql://localhost:53001/afcqztepbfci?user=postgres
[main] INFO org.flywaydb.core.internal.license.VersionPrinter - Flyway Community Edition 9.8.2 by Redgate
[main] INFO org.flywaydb.core.internal.license.VersionPrinter - See what's new here: https://flywaydb.org/documentation/learnmore/releaseNotes#9.8.2
[main] INFO org.flywaydb.core.internal.license.VersionPrinter - 
[main] INFO org.flywaydb.core.internal.database.base.BaseDatabaseType - Database: jdbc:postgresql://localhost:53001/afcqztepbfci (PostgreSQL 13.9)
[main] INFO org.flywaydb.core.internal.command.DbValidate - Successfully validated 1 migration (execution time 00:00.016s)
[main] INFO org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory - Creating Schema History table "public"."flyway_schema_history" ...
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Current version of schema "public": << Empty Schema >>
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Migrating schema "public" to version "1 - migration"
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Successfully applied 1 migration to schema "public", now at version v1 (execution time 00:00.029s)
---- first test, end spin up database ----

---- first test, begin shut down database ----
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully dropped pre-schema database level objects (execution time 00:00.000s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully cleaned schema "public" (execution time 00:00.024s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully cleaned schema "public" (execution time 00:00.009s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully dropped post-schema database level objects (execution time 00:00.000s)
---- first test, end shut down database ----

---- second test, begin spin up database ----
[main] INFO de.softwareforge.testing.postgres.junit5.EmbeddedPgExtension - Connection to jdbc:postgresql://localhost:53001/mnorzqllqsnm?user=postgres
[main] INFO org.flywaydb.core.internal.database.base.BaseDatabaseType - Database: jdbc:postgresql://localhost:53001/mnorzqllqsnm (PostgreSQL 13.9)
[main] INFO org.flywaydb.core.internal.command.DbValidate - Successfully validated 1 migration (execution time 00:00.003s)
[main] INFO org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory - Creating Schema History table "public"."flyway_schema_history" ...
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Current version of schema "public": << Empty Schema >>
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Migrating schema "public" to version "1 - migration"
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Successfully applied 1 migration to schema "public", now at version v1 (execution time 00:00.019s)
---- second test, end spin up database ----

---- second test, begin shut down database ----
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully dropped pre-schema database level objects (execution time 00:00.000s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully cleaned schema "public" (execution time 00:00.023s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully cleaned schema "public" (execution time 00:00.008s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully dropped post-schema database level objects (execution time 00:00.000s)
---- second test, end shut down database ----

---- third test, begin spin up database ----
[main] INFO de.softwareforge.testing.postgres.junit5.EmbeddedPgExtension - Connection to jdbc:postgresql://localhost:53001/ncnhfvvdxubv?user=postgres
[main] INFO org.flywaydb.core.internal.database.base.BaseDatabaseType - Database: jdbc:postgresql://localhost:53001/ncnhfvvdxubv (PostgreSQL 13.9)
[main] INFO org.flywaydb.core.internal.command.DbValidate - Successfully validated 1 migration (execution time 00:00.002s)
[main] INFO org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory - Creating Schema History table "public"."flyway_schema_history" ...
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Current version of schema "public": << Empty Schema >>
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Migrating schema "public" to version "1 - migration"
[main] INFO org.flywaydb.core.internal.command.DbMigrate - Successfully applied 1 migration to schema "public", now at version v1 (execution time 00:00.017s)
---- third test, end spin up database ----

---- third test, begin shut down database ----
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully dropped pre-schema database level objects (execution time 00:00.000s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully cleaned schema "public" (execution time 00:00.022s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully cleaned schema "public" (execution time 00:00.009s)
[main] INFO org.flywaydb.core.internal.command.clean.CleanExecutor - Successfully dropped post-schema database level objects (execution time 00:00.000s)
---- third test, end shut down database ----

Process finished with exit code 0
```

So... works as designed? Admittedly, with pg-extension you get a new database for every test execution. A unit test would be very helpful here.

**Comment by broccolai** [comment]
I'll try and make an isolated test later, this code previously worked before updating to 3.34.1-SNAPSHOT

[test where I use the flyway initializer with H2](https://github.com/broccolai/tickets/blob/rewrite/everything/core/src/test/java/love/broccolai/tickets/core/service/DatabaseStorageServiceTest.java)

```
  love.broccolai.tickets.core.service.DatabaseStorageServiceTest ✔ selectTickets()
  love.broccolai.tickets.core.service.DatabaseStorageServiceTest ✘ findTickets()

    org.jdbi.v3.core.statement.UnableToCreateStatementException: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "tickets_ticket" not found (this database is empty); SQL statement:
    INSERT INTO tickets_ticket(`status`, `creator`, `date`, `message`)
    VALUES ('OPEN', ?, ?, ?) [42104-214] [statement:"INSERT INTO tickets_ticket(`status`, `creator`, `date`, `message`)
    VALUES ('OPEN', :creator, :date, :message)", arguments:{positional:{}, named:{date:2022-11-23T12:38:10Z,creator:2bb0cdce-c056-4ff6-b761-322e73f0c6ea,message:Test Message}, finder:[]}]
        at app//love.broccolai.tickets.core.service.DatabaseStorageServiceTest.findTickets(DatabaseStorageServiceTest.java:89)
    Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "tickets_ticket" not found (this database is empty); SQL statement:
    INSERT INTO tickets_ticket(`status`, `creator`, `date`, `message`)
    VALUES ('OPEN', ?, ?, ?) [42104-214]
        at app//love.broccolai.tickets.core.service.DatabaseStorageServiceTest.findTickets(DatabaseStorageServiceTest.java:89)

  love.broccolai.tickets.core.service.DatabaseStorageServiceTest ✘ saveTicket()

    org.jdbi.v3.core.statement.UnableToCreateStatementException: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "tickets_ticket" not found (this database is empty); SQL statement:
    INSERT INTO tickets_ticket(`status`, `creator`, `date`, `message`)
    VALUES ('OPEN', ?, ?, ?) [42104-214] [statement:"INSERT INTO tickets_ticket(`status`, `creator`, `date`, `message`)
    VALUES ('OPEN', :creator, :date, :message)", arguments:{positional:{}, named:{date:2022-11-23T12:38:10Z,creator:eb2bf4f6-0155-4ebf-9b13-00bb8a916501,message:Hello!}, finder:[]}]
        at app//love.broccolai.tickets.core.service.DatabaseStorageServiceTest.saveTicket(DatabaseStorageServiceTest.java:57)
    Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "tickets_ticket" not found (this database is empty); SQL statement:
    INSERT INTO tickets_ticket(`status`, `creator`, `date`, `message`)
    VALUES ('OPEN', ?, ?, ?) [42104-214]
        at app//love.broccolai.tickets.core.service.DatabaseStorageServiceTest.saveTicket(DatabaseStorageServiceTest.java:57)
```

I'm not very familiar with the jdbi codebase but if nothing sticks out to you I'll poke around and see if I can figure more out

**Comment by hgschmie** [comment]
Hi @broccolai,

Thank you for the test.  That clears things up for me. The JdbiExtension is not intended to be registered as a class level field (static) but an instance field. That is why the method is run only once for you (before the first test) and so the cleanup removes the data from the database. If you remove the `static` qualifier from the H2_EXTENSION field, it should work.

The more interesting thing is why this worked for you in 3.34.0. I guess it is because before, we did not actually enforce the "cleanAfter", which is what #2179 does. I needed to upgrade Flyway to 8.5 so make it work with our version of H2.

I will add some documentation to emphasize how the extension is supposed to work to clear this up.
